plugin/advisor/pg/advisor_index_key_number_limit.go

```diff
@@ -0,0 +1,129 @@
+}
+
+// IndexKeyNumberLimitAdvisor is the advisor checking for index key number limit.
+type IndexKeyNumberLimitAdvisor struct {
+}
+
+// Check checks for index key number limit.
+func (*IndexKeyNumberLimitAdvisor) Check(ctx advisor.Context, statement string) ([]advisor.Advice, error) {
+	stmts, errAdvice := parseStatement(statement)
+	if errAdvice != nil {
+		return errAdvice, nil
+	}
+
+	level, err := advisor.NewStatusBySQLReviewRuleLevel(ctx.Rule.Level)
+	if err != nil {
+		return nil, err
+	}
+
+	payload, err := advisor.UnmarshalNumberLimitRulePayload(ctx.Rule.Payload)
+	if err != nil {
+		return nil, err
+	}
+	checker := &indexKeyNumberLimitChecker{
+		level: level,
+		title: string(ctx.Rule.Type),
+		max:   payload.Number,
+	}
+
+	for _, stmt := range stmts {
+		ast.Walk(checker, stmt)
+	}
+
+	if len(checker.adviceList) == 0 {
+		checker.adviceList = append(checker.adviceList, advisor.Advice{
+			Status:  advisor.Success,
+			Code:    advisor.Ok,
+			Title:   "OK",
+			Content: "",
+		})
+	}
+	return checker.adviceList, nil
+}
+
+type indexKeyNumberLimitChecker struct {
+	adviceList []advisor.Advice
+	level      advisor.Status
+	title      string
+	max        int
+}
+
+func (checker *indexKeyNumberLimitChecker) getMetaDataList(in ast.Node) []*indexMetaData {
+	var res []*indexMetaData
+
+	switch node := in.(type) {
+	case *ast.CreateIndexStmt:
+		if checker.max > 0 && len(node.Index.KeyList) > checker.max {
+			res = append(res, &indexMetaData{
+				indexName: node.Index.Name,
+				tableName: node.Index.Table.Name,
+				line:      node.LastLine(),
+			})
+		}
+	case *ast.CreateTableStmt:
+		for _, constraint := range node.ConstraintList {
+			if checker.max > 0 && indexKeyNumber(constraint) > checker.max {
+				res = append(res, &indexMetaData{
+					indexName: constraint.Name,
+					tableName: node.Name.Name,
+					line:      constraint.LastLine(),
+				})
+			}
+		}
+	case *ast.AddConstraintStmt:
+		if checker.max > 0 && indexKeyNumber(node.Constraint) > checker.max {
+			res = append(res, &indexMetaData{
+				indexName: node.Constraint.Name,
+				tableName: node.Table.Name,
+				line:      node.LastLine(),
+			})
+		}
+	}
+
+	return res
+}
+
+func (checker *indexKeyNumberLimitChecker) Visit(node ast.Node) ast.Visitor {
+	indexDataList := checker.getMetaDataList(node)
+
+	for _, index := range indexDataList {
+		checker.adviceList = append(checker.adviceList, advisor.Advice{
+			Status:  checker.level,
+			Code:    advisor.IndexKeyNumberExceedsLimit,
+			Title:   checker.title,
+			Content: fmt.Sprintf("The number of index `%s` in table `%s` should be not greater than %d", index.indexName, index.tableName, checker.max),
+			Line:    index.line,
+		})
+	}
+	return checker
+}
+
+func indexKeyNumber(constraint *ast.ConstraintDef) int {
+	switch constraint.Type {
+	case ast.ConstraintTypePrimary,
+		ast.ConstraintTypeUniqueUsingIndex,
+		ast.ConstraintTypePrimaryUsingIndex,
```

Comment:
Maybe no need to check these two cases, because they don't actually create new indexes and the key length is zero.